src/process/mod.rs

```diff
@@ -675,34 +672,69 @@ impl FromStr for FDTarget {
 #[derive(Clone)]
 pub struct FDInfo {
     /// The file descriptor
-    pub fd: u32,
+    pub fd: i32,
     /// The permission bits for this FD
     ///
     /// **Note**: this field is only the owner read/write/execute bits.  All the other bits
     /// (include filetype bits) are masked out.  See also the `mode()` method.
-    pub mode: libc::mode_t,
+    pub mode: u16,
     pub target: FDTarget,
 }
 
 impl FDInfo {
     /// Gets a file descriptor from a raw fd
-    pub fn from_raw_fd(pid: pid_t, raw_fd: i32) -> ProcResult<Self> {
+    pub fn from_raw_fd(pid: i32, raw_fd: i32) -> ProcResult<Self> {
         Self::from_raw_fd_with_root("/proc", pid, raw_fd)
     }
 
     /// Gets a file descriptor from a raw fd based on a specified `/proc` path
-    pub fn from_raw_fd_with_root(root: impl AsRef<Path>, pid: pid_t, raw_fd: i32) -> ProcResult<Self> {
+    pub fn from_raw_fd_with_root(root: impl AsRef<Path>, pid: i32, raw_fd: i32) -> ProcResult<Self> {
         let path = root.as_ref().join(pid.to_string()).join("fd").join(raw_fd.to_string());
         let link = wrap_io_error!(path, read_link(&path))?;
         let md = wrap_io_error!(path, path.symlink_metadata())?;
         let link_os: &OsStr = link.as_ref();
         Ok(Self {
-            fd: raw_fd as u32,
-            mode: (md.st_mode() as libc::mode_t) & libc::S_IRWXU,
+            fd: raw_fd,
+            mode: ((md.mode() as RawMode) & Mode::RWXU.bits()) as u16,
             target: expect!(FDTarget::from_str(expect!(link_os.to_str()))),
         })
     }
 
+    /// Gets a file descriptor from a directory fd and a path relative to it.
+    ///
+    /// `base` is the path to the directory fd, and is used for error messages.
+    fn from_process_at<P: AsRef<Path>, Q: AsRef<Path>>(
+        base: P,
+        dirfd: BorrowedFd,
+        path: Q,
+        fd: i32,
+    ) -> ProcResult<Self> {
+        let p = path.as_ref();
+        let root = base.as_ref().join(p);
+        let file = wrap_io_error!(
+            root,
+            rustix::fs::openat(
+                dirfd,
+                p,
+                OFlags::NOFOLLOW | OFlags::PATH | OFlags::CLOEXEC,
+                Mode::empty()
+            )
+        )
+        .map(|fdfd| File::from_fd(fdfd.into()))?;
+        let fdfd = file.as_fd();
```

Comment:
Why does this have to go through `File::from_fd()`?   Can't just used the `OwnedFd` returned by `openat`, and borrow it when calling `statat` and `readlinkat` ?

Reply:
Ah, right, we don't need a `FIle` here. That makes this code much simpler.